nuxeo-runtime/nuxeo-runtime-tomcat-adapter/src/main/java/org/nuxeo/runtime/tomcat/NuxeoDeployer.java

```diff
@@ -50,54 +54,70 @@ public String getHome() {
+            preprocess(container);
+        } else if (lifecycle instanceof StandardHost && Lifecycle.AFTER_START_EVENT.equals(type)) {
+            StandardHost container = (StandardHost) lifecycle;
+            checkFailures(container);
         }
     }
 
-    protected void handleEvent(ContainerBase container, LifecycleEvent event) {
+    protected void checkFailures(StandardHost host) {
+        boolean ok = Stream.of(host.findChildren()).map(Lifecycle::getState).allMatch(s -> s == LifecycleState.STARTED);
+        boolean strict = System.getProperty("nuxeo.start.strict") != null;
+
+        if (!ok && strict) {
+            // Throws an exception and let Tomcat to handle it via
+            // Catalina's shutdown hook. Otherwise, we could call
+            // `#stop()` and `#destroy()` by ourselves, but that means
+            // we're making a shutdown in AFTER_STARTED lifecycle
+            // event. It would be misleading.
+            throw new IllegalStateException("There're context(s) failed to start.");
+        }
+    }
+
+    protected void preprocess(Container container) {
         try {
             ClassLoader parentCl = container.getParentClassLoader();
-            String type = event.getType();
-            if (type == Lifecycle.BEFORE_START_EVENT) {
-                File homeDir = resolveHomeDirectory();
-                File bundles = new File(homeDir, "bundles");
-                File lib = new File(homeDir, "lib");
-                File deployerJar = FrameworkBootstrap.findFileStartingWidth(bundles, "nuxeo-runtime-deploy");
-                File commonJar = FrameworkBootstrap.findFileStartingWidth(bundles, "nuxeo-common");
-                if (deployerJar == null || commonJar == null) {
-                    System.out.println("Deployer and/or common JAR (nuxeo-runtime-deploy* | nuxeo-common*) not found in "
-                            + bundles);
-                    return;
-                }
-                ArrayList<URL> urls = new ArrayList<URL>();
-                File[] files = lib.listFiles();
-                if (files != null) {
-                    for (File f : files) {
-                        if (f.getPath().endsWith(".jar")) {
-                            urls.add(f.toURI().toURL());
-                        }
+            File homeDir = resolveHomeDirectory();
+            File bundles = new File(homeDir, "bundles");
+            File lib = new File(homeDir, "lib");
+            File deployerJar = FrameworkBootstrap.findFileStartingWidth(bundles, "nuxeo-runtime-deploy");
+            File commonJar = FrameworkBootstrap.findFileStartingWidth(bundles, "nuxeo-common");
+            if (deployerJar == null || commonJar == null) {
+                System.out.println("Deployer and/or common JAR (nuxeo-runtime-deploy* | nuxeo-common*) not found in "
+                        + bundles);
+                return;
+            }
+            ArrayList<URL> urls = new ArrayList<URL>();
```

Comment:
Could you replace type by `List` ?

Reply:
This was just moving code around so this could be a separate cleanup (diff -b will show this unchanged).

Reply:
This cleanup (moving the code + refactoring) is done via commit 8859020 and its fix-up fc19295. I'll squash them together before merging. Is it what you want me to do, Florent?

Reply:
No, don't squash formatting with other code changes. Commits must be readable by themselves.